### ISSUE TITLE
Improve "Watch" Mode For Truffle CLI

### ISSUE THREAD

**Issue by JimLynchCodes** [trufflesuite/truffle#4084]
## Issue

I am trying to run the tests automatically so I can do TDD with an easy, fun worklfow.

However, the watch mode doesn't seem to be working properly (not running tests) when I do this command:
```
truffle watch
```

or

```
truffle watch --tests
```

Also, it's making me nervous and sad that the docs are saying the watch command in deprecated. 😢

![Screen Shot 2021-05-31 at 11 32 56 PM](https://user-images.githubusercontent.com/5354163/120262849-89d8b480-c268-11eb-82e9-383a39338727.png)

Also, it's odd to me that there is no recommendation for the "new way" to run tests in watch mode with this now deprecated. It's kind of just like, "welp, sorry but you're beat!" 😅

---

Honestly though, I think watch mode as a flag for "truffle test" would be the most natural and intuitive way to have it.

Consider other javascript test runners like jest or mocha:

```
jest --watch
```

```
mocha --watch
```

with `-w` alis:
```
mocha -w
```

It would be great if truffle test (and possibly other truffle commands) supported things syntax:

```
truffle test --watch
```

or with the alias:
```
truffle test -w
```

Thanks! ♥️

**Comment by eggplantzzz** [comment]
Sorry yes, this command will be deprecated in the future. We don't think that this is really in the scope of what Truffle should do and instead think that a task running tool like gulp or grunt would be more appropriate for this kind of thing. Perhaps that should be added to the docs on the website (recommending grunt or gulp)?

**Comment by fainashalts** [comment]
Would be neat to add watch functionality as a flag to existing commands, even if we deprecate `truffle watch`. We'll add this to the backlog for consideration. PRs welcome in the meantime! Thanks @JimLynchCodes!

**Comment by JimLynchCodes** [comment]
@eggplantzzz  I am not sure how exactly the --watch flag should be implemented, but I don't think the user really cares what it uses under the hood as long as it works! 😁

**Comment by JimLynchCodes** [comment]
truffle is just wrapping mocha isn't it? 

can't the flag just be passed down to the mocha process since it's already supported there?

**Comment by JimLynchCodes** [comment]
Thanks @fainashalts! It would be greater to get this as a higher priority thing so my friends and I can start using it! 😉

In the meantime, any links for doing this in a truffle project with gulp or nodemon or something would be appreciated!

**Comment by JimLynchCodes** [comment]
Here's a nice way of doing it that I've been using:

(make you have a package.json file first: `npm init`)

Then install nodemon as a dev dependency:
```
npm i -D nodemon
```

Use these scripts in your package.json for running tests and allowing it to watch solidity and javascript files, running "truffle test" when there are any changes:

```
"scripts": {
    "test": "truffle test",
    "test:watch": "nodemon --ext 'js,sol' --exec 'truffle test'"
  },
```

**Comment by betochf** [comment]
@JimLynchCodes soooo, you're saying that the flag should execute the process with mocha instead of truffle?
I'd like to try to pull the enhancement, so I'm looking for any ideas.

**Comment by eggplantzzz** [comment]
@JimLynchCodes Yes, if you include a `mocha` property in your `truffle-config.js` that stuff should get passed through to mocha. I'm not sure if command line flags will though.

**Comment by sahilrajput03** [comment]
mann... I just came here to find how to make watch mode work and I am disappointed by this thread here...

Watch mode is really the key feature for any development environment..., and I might do a PR in future when I get time to add this watch mode feature. I like hardhat specifically for this feature though and I want truffle to support this as well.
